builtin/mainmenu/settings/dlg_settings.lua

```diff
@@ -0,0 +1,514 @@
+			"image_button[0.75,0;0.75,0.75;", core.formspec_escape(defaulttexturedir .. "clear.png"), ";search_clear;]",
+			"tooltip[search;", fgettext("Search"), "]",
+			"tooltip[search_clear;", fgettext("Clear"), "]",
+		"container_end[]",
+		"scroll_container[0.25,1.25;", tostring(left_pane_width), ",",
+				tostring(tabsize.height - 1.5), ";leftscroll;vertical;0.1]",
+		"style_type[button;border=false;bgcolor=#3333]",
+		"style_type[button:hover;border=false;bgcolor=#6663]",
+	}
+
+	local y = 0
+	local last_section = nil
+	for _, other_page in ipairs(filtered_pages) do
+		if other_page.section ~= last_section then
+			fs[#fs + 1] = "style_type[label;textcolor="
+			fs[#fs + 1] = "#ff0"
+			fs[#fs + 1] = "]"
+			fs[#fs + 1] = ("label[0.1,%f;%s]"):format(y + 0.41, fgettext(other_page.section))
+			last_section = other_page.section
+			fs[#fs + 1] = "style_type[label;textcolor=]"
+			y = y + 0.82
+		end
+		fs[#fs + 1] = ("box[0,%f;%f,0.8;%s]"):format(
+			y, left_pane_width, other_page.id == page_id and "#467832FF" or "#3339")
+		fs[#fs + 1] = ("button[0,%f;%f,0.8;page_%s;%s]")
+			:format(y, left_pane_width, other_page.id, fgettext(other_page.title))
+		y = y + 0.82
+	end
+
+	if #filtered_pages == 0 then
+		fs[#fs + 1] = "label[0.1,0.1;"
+		fs[#fs + 1] = fgettext("No results")
+		fs[#fs + 1] = "]"
+	end
+
+	fs[#fs + 1] = "scroll_container_end[]"
+
+	if y >= tabsize.height - 1.25 then
+		fs[#fs + 1] = make_scrollbaroptions_for_scroll_container(tabsize.height - 1.5, y, 0.1)
+		fs[#fs + 1] = ("scrollbar[%f,1.25;%f,%f;vertical;leftscroll;%f]"):format(
+				left_pane_width + 0.25, scrollbar_w, tabsize.height - 1.5, dialogdata.leftscroll or 0)
+	end
+
+	fs[#fs + 1] = "style_type[button;border=;bgcolor=]"
+
+	if not dialogdata.components then
+		dialogdata.components = page and build_page_components(page) or {}
+	end
+
+	local right_pane_width = tabsize.width - left_pane_width - 0.375 - 2*scrollbar_w - 0.25
+	fs[#fs + 1] = ("scroll_container[%f,0;%f,%f;rightscroll;vertical;0.1]"):format(
+			tabsize.width - right_pane_width - scrollbar_w, right_pane_width, tabsize.height)
+
+	y = 0.25
+	for i, comp in ipairs(dialogdata.components) do
+		fs[#fs + 1] = ("container[0,%f]"):format(y)
+
+		local avail_w = right_pane_width - 0.25
+		if not comp.full_width then
+			avail_w = avail_w - 1.4
+		end
+		if comp.max_w then
+			avail_w = math.min(avail_w, comp.max_w)
+		end
+
+		local comp_fs, used_h = comp:get_formspec(avail_w)
+		fs[#fs + 1] = comp_fs
+
+		fs[#fs + 1] = "style_type[image_button;border=false;padding=]"
+
+		if comp.changed and comp.setting and comp.setting.default then
+			fs[#fs + 1] = ("image_button[%f,%f;0.5,0.5;%s;%s;]"):format(
+					right_pane_width - 1.4, used_h / 2 - 0.25, reset_icon_path, "reset_" .. i)
+			fs[#fs + 1] = ("tooltip[%s;%s]"):format(
+					"reset_" .. i, fgettext("Reset setting to default ($1)", tostring(comp.setting.default)))
+		end
+
+		if comp.info_text and comp.info_text ~= "" then
+			used_h = math.max(used_h or 0, 0.5)
```

Comment:
This implies that `used_h` may be nil, but `used_h` is used before this line.  Can it actually be nil?  If so, this should be moved up.